lib/puppet/parser/functions/clamps_hiera_defaults.rb

```diff
@@ -0,0 +1,50 @@
+#!/opt/puppet/bin/ruby
+require 'puppet/face'
+require 'yaml'
+
+module Puppet::Parser::Functions
+  newfunction(:clamps_hiera_defaults, :type => :rvalue, :arity => 0, :doc => <<-EOS
+  This function reads through the modules on disk via the resource_type face
+  It then determines the parameters for these classes and their defaults.
+  It creates hiera data binding compatible key names i.e. 'apache::user'
+  and the evaluated ruby object as the value. This preserves boolean/hash/array 
+  EOS
+) do |args|
+   
+  #Puppet.parse_config
+  environment = Puppet[:environment] || 'production'
+  resources = Puppet::Face[:resource_type, '0.0.1'].search('.*',{:extra => { 'environment' => environment }}) 
+   
+  response = Hash.new
+    resources.each do |resource|
+     resource.arguments.each do |k,v|
+        # Walk the AST types and evaluate them, Puppet 4 may need updates
+        case v.class.to_s
+          when "Puppet::Parser::AST::String"
+            result = v.evaluate('subscope')
+          when "Puppet::Parser::AST::Undef"
+            # "We don't want undef values..."
+            next 
+          #when "Puppet::Parser::AST::Variable"
+          #  # Remove the $ in the name and wrap in hiera
+          #  # inpolation to defer the lookup until runtime
+          #  result = "%{#{v.to_s[1..-1]}}"
```

Comment:
Its coming back as soon as I figure out the best way to call `.to_yaml` with a key with colons in the name. Basically `.to_yaml` automatically double quotes any key or value with `:` in the name and there is no option to turn it off. You then end up with:

```yaml
"apache::user": "%{i::am::a::var}"
```
However if `$i::am::a::var` is boolean the quotes cause it to be stringified. Ideally I would write the value out as:

```yaml
"apache::user": %{i::am::a::var}
```

but '.to_yaml' is overly paranoid about `:`

git does not really help here as my intial commit was it quoted 
https://github.com/puppetlabs/clamps/commit/b9f0ec91f7e811a4c9ba29d8219c454decb3b029#diff-f5b2159f2aed2f632e7c1b954df24875R28